src/object.d

```diff
@@ -2975,8 +2975,11 @@ Key[] keys(T : Value[Key], Value, Key)(T aa) @property
         alias realAA = aa;
     else
         const(Value[Key]) realAA = aa;
-    auto a = cast(void[])_aaKeys(*cast(inout(AA)*)&realAA, Key.sizeof, typeid(Key[]));
-    auto res = *cast(Key[]*)&a;
+    typeof(return) res;
+    () @trusted {
+        auto a = cast(void[])_aaKeys(*cast(inout(AA)*)&realAA, Key.sizeof, typeid(Key[]));
+        res = *cast(Key[]*)&a;
+    } ();
     static if (__traits(hasPostblit, Key))
```

Comment:
I don't think copy constructors are supported yet, right @RazvanN7?

Reply:
As both of the asserts below fail I presume that AAs doesnt' work with copy constructors.

@trusted pure unittest {
    size_t kc;                     // key copy count
    size_t vc;                     // value copy count

    @safe struct Key {
        this(ref return scope typeof(this) rhs) { kc++; }
        string _;
    }

    @safe struct Value {
        this(ref return scope typeof(this) rhs) { vc++; }
        string _;
    }

    alias AA = Value[Key];
    AA aa = [ Key("x") : Value("x") ];

    assert(kc == 0);
    const keys = aa.keys;
    assert(kc == 1);

    assert(vc == 0);
    const values = aa.values;
    assert(vc == 1);
}

What would be the correct way to correct keys and values using `__traits(hasCopyConstructor, ...)`?

Reply:
You can take a look at https://github.com/dlang/druntime/pull/3239 that does this for static array construction.

Reply:
I'd prefer postponing that for another MR.